### PR TITLE
Merge branch rlos2021_cfe to master

### PR DESCRIPTION
**Pull request by JuiP**
Creating this PR to check we haven't broken anything and the CI tests pass.

### CODE REVIEW

File: estimators/bandits/base.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+""" Interface for implementation of contextual bandit estimators """
+
+from abc import ABC, abstractmethod
+from typing import List
+
+class Estimator(ABC):
+	""" Interface for implementation of contextual bandit estimators """
+
+	@abstractmethod
+	def add_example(self, p_log: float, r: float, p_pred: float, count: float) -> None:
```
Comment (jackgerrits) [inline estimators/bandits/base.py:10]: Consider renaming these to be more descriptive, `prob_log`/`prob_online`, `reward`, `prob_predicted`/`prob_offline`

File: estimators/bandits/clopper_pearson.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        #
+        # 'n':   IPS of numerator
+        # 'N':   total number of samples in bin from log (IPS = n/N)
+        # 'c':   max abs. value of numerator's items (needed for Clopper-Pearson confidence intervals)
+        #
+        #################################################################################################
+
+        self.data = {'n':0.,'N':0,'c':0.}
+
+    def add_example(self, p_log, r, p_pred, count=1):
```
Comment (jackgerrits) [inline estimators/bandits/clopper_pearson.py:18]: types and count is a float so should be 1.0

File: estimators/bandits/clopper_pearson.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        self.data['N'] += count
+        if p_pred > 0:
+            p_over_p = p_pred/p_log
+            if r != 0:
+                self.data['n'] += r*p_over_p*count
+                self.data['c'] = max(self.data['c'], r*p_over_p)
+
+    def get(self, alpha=0.05):
```
Comment (jackgerrits) [inline estimators/bandits/clopper_pearson.py:26]: types

File: estimators/bandits/clopper_pearson.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        bounds = []
+        num = self.data['n']
+        den = self.data['N']
+        maxWeightedCost = self.data['c']
```
Comment (jackgerrits) [inline estimators/bandits/clopper_pearson.py:30]: should use snake case

File: estimators/bandits/cressieread.py [hunk 5]
```diff
@@ -24,7 +25,7 @@ def add_example(self, p_log, r, p_pred, count=1):
             self.wmax = max(self.wmax, w)
             self.wmin = min(self.wmin, w)
 
-    def get_estimate(self, rmin=0, rmax=1):
+    def get(self):
```
Comment (jackgerrits) [inline estimators/bandits/cressieread.py:28]: types

File: estimators/bandits/cressieread.py [hunk 6]
```diff
@@ -53,17 +54,20 @@ def get_estimate(self, rmin=0, rmax=1):
...
+class Interval(base.Interval):
     # NB: This works better you use the true wmin and wmax
     #     which is _not_ the empirical minimum and maximum
     #     but rather the actual smallest and largest possible values
-    def __init__(self, wmin=0, wmax=inf):
+    def __init__(self, wmin=0, wmax=inf, rmin=0, rmax=1):
         assert wmin < 1
         assert wmax > 1
 
         self.wmin = wmin
```
Comment (jackgerrits) [inline estimators/bandits/cressieread.py:65]: can you add type annotations for these class fields

File: estimators/bandits/gaussian.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+            p_over_p = p_pred/p_log
+            if r != 0:
+                self.data['n'] += r*p_over_p*count
+                self.data['SoS'] += ((r*p_over_p)**2)*count
+
+    def get(self, alpha=0.05):
+        bounds = []
+        num = self.data['n']
+        den = self.data['N']
+        SoS = self.data['SoS']
```
Comment (jackgerrits) [inline estimators/bandits/gaussian.py:30]: longer name with snake case is preferable

File: estimators/bandits/gaussian.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+        if SoS > 0.0 and den > 1:
+            zGaussianCdf = stats.norm.ppf(1-alpha/2)
```
Comment (jackgerrits) [inline estimators/bandits/gaussian.py:33]: snake case

File: estimators/ccb/base.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
+""" Interface for implementation of conditional contextual bandits estimators """
+
+from abc import ABC, abstractmethod
+from typing import List
+
+class Estimator(ABC):
+	""" Interface for implementation of conditional contextual bandits estimators """
+
+	@abstractmethod
+	def add_example(self, p_log: List, r: List, p_pred: List, count: float) -> None:
```
Comment (jackgerrits) [inline estimators/ccb/base.py:10]: Can you make these typed lists? `List[float]`

File: estimators/ccb/first_slot.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+class Estimator(base.Estimator):
+    def __init__(self, bandits_estimator):
+        self.estimator = bandits_estimator
+
+    def add_example(self, p_logs, r, p_preds, count=1):
+        """Expects lists for logged probabilities, rewards and predicted probabilities. These should correspond to each slot."""
+
+        if not isinstance(p_logs, list) and not isinstance(r, list) and not isinstance(p_preds, list):
+            raise('Error: p_logs, r and p_preds must be lists')
```
Comment (jackgerrits) [inline estimators/ccb/first_slot.py:11]: Can you raise a specific error type such as `ValueError`

File: estimators/bandits/clopper_pearson.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
+import math
+from scipy.stats import beta
+from estimators.bandits import base
+
+class Interval(base.Interval):
+
+    def __init__(self):
+        ################################# Aggregates quantities #########################################
+        #
+        # 'n':   IPS of numerator
```
[the remark this answers is not in the document]
Comment (ataymano) [reply estimators/bandits/clopper_pearson.py:10]: I agree, but let me fix it in next PR.
This PR is mostly about restructuring the project, changing interfaces and adding tests and not supposed to touch implementation much.

File: estimators/bandits/base.py (hunk 1 above)
Comment (ataymano) [reply estimators/bandits/base.py:10]: It sounds like big change which require to touch a lot inside implementations - let me work on it in one of next PRs

**Review by jackgerrits — APPROVED** [review]

**Review by olgavrou — APPROVED** [review]

**Comment by ataymano** [comment]
@marco-rossi29, @pmineiro, please feel free to provide feedback if you have any - happy to work on it in next PRs.

**Pull request merged**